### PR TITLE
Allow Emoji Characters in the URL

### PR DESCRIPTION
**Pull request by AntonNguyen**
Emoji characters are a valid in a URL, so the validator should be updated to allow this edge case to pass.

Fun Emoji/Browser facts
1. Browsers utilize something called `punycoder` to convert something like `🍕` to `xn--vi8h`
2. At the moment, the only ccTLD that supports emojis in the domain name, is `.ws`!

refs #8368

### CODE REVIEW

**Comment by burzum** [comment]
Looks good. But could we add a `@link` doc block to some RFC or spec for the emoji stuff? I haven't been aware that there is something like that and also not that you're allowed to use them in URLs.

File: src/Validation/Validation.php [hunk 1]
```diff
@@ -908,7 +908,9 @@ public static function range($check, $lower = null, $upper = null)
     public static function url($check, $strict = false)
     {
         static::_populateIp();
-        $alpha = '0-9\p{L}\p{N}';
+
+        $emoji = '\x{1F300}-\x{1F6FF}';
```
Comment (markstory) [inline src/Validation/Validation.php]: Should we go higher than U+1F6FF? The [emoji table on wikipedia](https://en.wikipedia.org/wiki/Emoji) has 🧀 at U+1F9C0. The points above 1F6FF are a bit sparse but there are a few good ones up there.

Comment (AntonNguyen) [reply src/Validation/Validation.php]: Absolutely!

File: src/Validation/Validation.php [hunk 2]
```diff
@@ -904,11 +904,14 @@ public static function range($check, $lower = null, $upper = null)
...
      * @return bool Success
+     * @link https://tools.ietf.org/html/rfc3986
      */
     public static function url($check, $strict = false)
     {
         static::_populateIp();
-        $alpha = '0-9\p{L}\p{N}';
+
+        $emoji = '\x{1F190}-\x{1F9EF}';
+        $alpha = '0-9(\p{L}\p{N}' . $emoji;
```
Comment (ADmad) [inline src/Validation/Validation.php]: Is the addition of `(` in the regex intentional? It wasn't there earlier.

**Comment by saeideng** [comment]
maybe `urlWithEmoji()`  instead of `url()`

**Comment by markstory** [comment]
@saeideng Why a separate method?

**Comment by saeideng** [comment]
maybe someone unlike the emoji in Url

**Comment by markstory** [comment]
Someone not liking emoji in urls doesn't make those URLs less valid though.

File: src/Validation/Validation.php (hunk 2 above)
Comment (AntonNguyen) [reply src/Validation/Validation.php]: Nice catch!

**Comment by inoas** [comment]
It would be great if the validator allowed validation of:
- asciiOnlyUrl
- utf8UrlNoEmoji
- utf8Url

... or similar.

**Comment by markstory** [comment]
@inoas I don't think we need to meet all those use cases in the validators that CakePHP provides. If people need to support a subset of valid URLs they can create their own patterns.

**Comment by inoas** [comment]
Well it may be a valid URL but that doesn't mean it is a good URL and the validator before allowing emojis > the validator after allowing them (also it is not BC technically).

**Comment by dereuromark** [comment]
I am with @markstory on this one:
If the current http specs allow for these URLs, and those are technically valid/correct, then this warrants a fix in a patch release.

If you don't like them, feel free to either use your own validation rule for URLs or simply add a validation rule that sniffs for the '\x{1F190}-\x{1F9EF}' 4 byte range and disallows their use. I don't think this is needed or justified, however, in the majority of use cases.

File: src/Validation/Validation.php (hunk 2 above)
Comment (markstory) [reply src/Validation/Validation.php]: I can fix that before merging.

**Comment by markstory** [comment]
Thanks!

**Pull request merged**